arrow-json/src/raw/converter.rs

```
@@ -0,0 +1,388 @@
+            TapeElement::Null => Ok(None),
+            TapeElement::True => Ok(NumCast::from(1)),
+            TapeElement::False => Ok(NumCast::from(0)),
+            TapeElement::String(idx) => {
+                let s = tape.get_string(idx);
+                let value = P::parse(s).ok_or_else(|| {
+                    ArrowError::JsonError(format!(
+                        "failed to parse \"{s}\" as {}",
+                        data_type
+                    ))
+                })?;
+
+                Ok(Some(value))
+            }
+            TapeElement::Number(idx) => {
+                parse_number::<P>(tape.get_string(idx), data_type)
+            }
+            d => Err(tape_error(d, "primitive")),
+        }
+    }
+
+    #[inline]
+    fn to_string<'a>(
+        &self,
+        tape: &'a Tape<'_, Self>,
+        idx: u32,
+    ) -> Result<Option<Cow<'a, str>>, ArrowError>
+    where
+        Self: Sized,
+    {
+        match tape.get(idx) {
+            TapeElement::Null => Ok(None),
+            TapeElement::True => Ok(Some("true".into())),
+            TapeElement::False => Ok(Some("false".into())),
+            TapeElement::String(idx) | TapeElement::Number(idx) => {
+                Ok(Some(tape.get_string(idx).into()))
+            }
+            TapeElement::StartList(_) => Ok(Some(obj_to_json_string(tape, idx)?)),
+            TapeElement::StartObject(_) => Ok(Some(obj_to_json_string(tape, idx)?)),
+            d => Err(tape_error(d, "string")),
+        }
+    }
+}
+
+fn obj_to_json_string<'a, C: TapeConverter>(
+    tape: &'a Tape<'_, C>,
+    idx: u32,
+) -> Result<Cow<'a, str>, ArrowError> {
+    match tape.get(idx) {
+        TapeElement::Null => Ok("null".into()),
+        TapeElement::True => Ok("true".into()),
+        TapeElement::False => Ok("false".into()),
+        TapeElement::Number(idx) => Ok(tape.get_string(idx).into()),
+        TapeElement::String(idx) => Ok(format!("\"{}\"", tape.get_string(idx)).into()),
```

I'm not sure this will correctly escape newlines or quotes?

---

Good catch. Do you think using serde_json is a good approach to solve this?

---

To be completely honest I struggle to understand the use-case, the performance will be bad and it seems like a rather strange API... This sort of schema coercion I'm not sure really belongs in the reader

---

The use case is that we don't always control the json format we're reading from, but we can control the Schema.

And it's the behaviour given by infer_json_schema:
```
        let schema = Schema::new(vec![
            Field::new("c1", DataType::Utf8, true),
            Field::new("c2", DataType::Utf8, true),
        ]);

        let inferred_schema = infer_json_schema_from_iterator(
            vec![
                Ok(serde_json::json!({"c1": 1, "c2": 1})),
                Ok(serde_json::json!({"c1": "a", "c2": 0})),
                Ok(serde_json::json!({"c1": true, "c2": "ok"})),
            ]
            .into_iter(),
        )
        .unwrap();

        assert_eq!(inferred_schema, schema);
```

---

> And it's the behaviour given by infer_json_schema:

infer_json_schema coerces primitives, I'm not aware of it coercing objects or lists to strings? If it is that is not intentional

---

Oh, no, it's just for primitives, and only to strings, if I'm not mistaken. Let me see what it coerces between number and bool.

I probably just got creative there and asked why not? But thinking about it, what I really need is to be able to read json with the schema that was inferred before, and use it to save to Parquet.

Maybe we don't even need a flag for it, shouldn't this be the default behaviour since infer_json_schema does it by default?

---

Yes, confirmed that if you mismatch numbers and bools, it will also consider them to be strings.

---

> since infer_json_schema does it by default

I think having an option that disables strict enforcement along the lines of the schema inference makes sense :+1:, I do think it should still be optional as some workflows care about schema consistency and would rather fail loud :smile:

---

Then I'll keep it very simple, then. Just a bool in a struct for now.